### PR TITLE
Fix and improve Enum.sum/1 for ranges

### PR DESCRIPTION
**Pull request by eksperimental**
Lastest changes introduced in Enum.sum in #4414 changed the type of the return value
(from integer to float) when argument was a range. This commit fixes this.

Adds a spec when dealing with Range.t

It simplifies the function by making it more recursive,
and also deals with ranges where first and last are equal

Tests added

/cc @Qqwy

### CODE REVIEW

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -1413,15 +1413,18 @@ defmodule Enum do
...
 
   """
+  @spec sum(Range.t) :: integer
   @spec sum(t) :: number
   def sum(enumerable)
 
-  def sum(first..last) when last > first do
-    ((last * (last + 1)) / 2) - ((first * (first - 1)) / 2)
-  end
+  def sum(first..last) when first == last,
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: We can write it as `sum(first..first)`.

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: oh, yeah.. I knew i was missing something in that line! thank you @lexmag

**Comment by josevalim** [comment]
Instead of using trunc, we should use

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -1413,15 +1413,18 @@ defmodule Enum do
...
+    do: first
 
-  def sum(first..last) do
-    ((first * (first + 1)) / 2) - ((last * (last - 1)) / 2)
+  def sum(first..last) when last < first,
+    do: sum(last..first)
+
+  def sum(first..last) when last > first do
+    trunc( ((last * (last + 1)) - (first * (first - 1))) / 2 )
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Instead of `trunc` we probably can use `div`.

**Comment by josevalim** [comment]
Instead of using `trunc` + `/`, we should simply use `div`. :)

File: lib/elixir/lib/enum.ex (hunk 2 above)
Comment (whatyouhide) [inline lib/elixir/lib/enum.ex]: (also, no spaces around parens :smiley:)

File: lib/elixir/test/elixir/enum_test.exs [hunk 3]
```diff
@@ -428,6 +428,11 @@ defmodule EnumTest do
     assert Enum.sum([1]) == 1
     assert Enum.sum([1, 2, 3]) == 6
     assert Enum.sum([1.1, 2.2, 3.3]) == 6.6
+    assert Enum.sum([-3, -2, -1, 0, 1, 2, 3]) == 0
```
Comment (whatyouhide) [inline lib/elixir/test/elixir/enum_test.exs]: Is this related to this PR?

Comment (eksperimental) [inline lib/elixir/test/elixir/enum_test.exs]: not per se, but I felt adding a test that deals with negative numbers was useful. should i remove it?

Comment (whatyouhide) [inline lib/elixir/test/elixir/enum_test.exs]: My personal opinion is probably yes, but maybe it's ok since this is a small PR. Let's have @lexmag or @josevalim decide :)

**Comment by eksperimental** [comment]
all suggestions pushed,
we just need to decide what to do with the test https://github.com/elixir-lang/elixir/pull/4484#discussion-diff-58778530R431

File: lib/elixir/lib/enum.ex [hunk 4]
```diff
@@ -1413,15 +1413,18 @@ defmodule Enum do
       6
 
   """
+  @spec sum(Range.t) :: integer
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: We can't have more than one `@spec` for a function.

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: I thought we could https://github.com/elixir-lang/ex_doc/issues/519

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: if that's not possible, what is the proper way to convey that `@spec`?

Comment (lexmag) [inline lib/elixir/lib/enum.ex]: Unfortunately we can't, it brings a dialyzer warning.
Tereis one of the fix commits resulted from the linked issue: https://github.com/elixir-lang/elixir/pull/4432/commits/ee66a9d1096f29c513878ecac8162684b6a1f533.

Comment (lexmag) [inline lib/elixir/lib/enum.ex]: We should stick with `@spec sum(t) :: number` for now.

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: thank you @lexmag, I noticed of that right after I posted my question.

then, my question is... shouldn't Elixir generate a warning when there are overlapping specs then? or it is something that we shouldn't worry about since it's valid Elixir code and just Dialyzer doesn't like it?

Comment (whatyouhide) [inline lib/elixir/lib/enum.ex]: Probably the latter :smile:

Comment (lexmag) [inline lib/elixir/lib/enum.ex]: I think for now it is Dialyzer's responsibility.

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: anybody knows if Dialyzer has got a public repository where to post this issue?

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: I see it's part of OTP

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: lib/elixir/lib/enum.ex [hunk 5]
```diff
@@ -1416,12 +1416,14 @@ defmodule Enum do
...
-  end
+  def sum(first..first),
+    do: first
 
-  def sum(first..last) do
-    ((first * (first + 1)) / 2) - ((last * (last - 1)) / 2)
+  def sum(first..last) when last < first,
+    do: sum(last..first)
+
+  def sum(first..last) when last > first do
```
Comment (TBK145) [inline lib/elixir/lib/enum.ex]: Isn't this when statement unnecessary because it is already implied by the previous function heads?

Comment (whatyouhide) [inline lib/elixir/lib/enum.ex]: Yep, not needed. It may aid reading somehow, not sure though :)

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: I intentionally left it for clarity

Comment (michalmuskala) [inline lib/elixir/lib/enum.ex]: What happens with the case `first == last`?

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: Is there a style for ommitting the guard in the second clause? I don't mind skipping it if that's the standard or you guys thing it's better that way

Comment (whatyouhide) [inline lib/elixir/lib/enum.ex]: @eksperimental leave it there for now, so we don't add a commit with more changes to something that works well this way. This is my opinion of course :)

**Comment by Qqwy** [comment]
Thanks a lot for fixing this :smiley: !

As for keeping vs. omitting unnecessary guards: I believe that the style convention is to omit them, as usually it is clear enough from the context that it will only be called when the earlier clauses did not match.

File: lib/elixir/lib/enum.ex (hunk 5 above)
Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: @whatyouhide I will see if I find some other clauses that can be cleaned, i will put them all together in one commit.
thanks everyone for all comments and contributions
